### ISSUE TITLE
What to do about :noswapfile

### ISSUE THREAD

**Issue by urticadioica** [igrigorik/vimgolf#136]
Vim 7.4.213 introduced the `:noswapfile` command, which has no real use in vimgolf, but had the side effect of stealing the expansion of `:no<S-Tab>`, which used to expand to `:normal`. The `:normal` command is very common in optimal solutions, and `:norm` is a stroke longer than `:no<S-Tab>`.

We've always been on shaky ground using command expansions in vimgolf. Vim sometimes adds extra commands or functions and messes things up. Once they're there, they're not going away. It would be awkward to ban by rule though. Very few command expansions are tactically important anyway.

We should decide what to do about it. I see 3 reasonable ideas:
### Require old Vims for best stroke counts

This is effectively what we're doing now. It's hard to expect every user to use the same version of Vim, so there's nothing stopping players from downgrading or keeping multiple Vims around to exploit different behaviors. Kind of ugly. Involves a lot of "Requires Vim <= X.X.XXX" comments. But it takes no work to implement.

This isn't the first version difference that's come up. It's just the biggest. We'll be doing this anyway.
### `<S-Tab>` is 2 strokes

I've wanted to do this with cursor keys anyway. Doing the same to `<S-Tab>` would create stroke parity between `:norm` and `:no<S-Tab>`. It would also leave almost no tactical excuse to use non-ascii commands, which is attractive in a way.

There's an ergonomic point to be made too. There's no way `:no<S-Tab>` is faster to type than `:norm`.

People have made solutions counting on the fact that `:no<S-Tab>` saves a stroke, but beyond that, it's interestingly balanced. A stroke less and you'd use it everywhere; a stroke more and it'd be hardly used. Counting an extra stroke for `:no<S-Tab>` would buff ex mode, register macros, and various other strategies, while sharply reducing the use for `:normal`.

In short, I prefer the gameplay effects of the shorter `:no<S-Tab>`. While this option is attractive, I won't recommend it.
### Iron out the version difference in the vimrc

Here's the best I've come up with:

```
function _MuchSTab()
    call feedkeys("\<S-Tab>", 'tn')
    if getcmdline()[:getcmdpos()-2] =~ '\<no$'
        call feedkeys("\<S-Tab>", 'tn')
    endif
    return ''
endfunction

if exists(':noswapfile')
    cnoremap <expr> <S-Tab> _MuchSTab()
endif
```

Allowing minimal side effects is tricky. Maybe someone else can do better. But I think this is good enough to be better than the other options.

I just hope the vimrc doesn't end up full of this stuff.

**Comment by zuloloxi** [comment]
I created an issue for vim, here the link: https://code.google.com/p/vim/issues/detail?id=237&can=1&start=200, but was rejected, The patch modify the noswapfile command to no_swapfile.
Best regards !

**Comment by igrigorik** [comment]
Did we reach any conclusion on this one?

**Comment by urticadioica** [comment]
A few people still keep an old Vim around to save the stroke. (A newer feature, visual increment, is useful far more often, and has actually knocked out quite a few top `:normal`s.). Basically it's a similar situation to cursor keys: some people use them, some don't, so there's sometimes a dispute over what the true Top Score is. Vimgolf certainly works better when we resolve those disputes.

The only option I see now for fixing the problem is changing the stroke counting, making `<S-Tab>`, `<Left>`, etc., count as 2 "strokes". It's the option I'd like to take, but it has a few caveats, being a little more complex and something that needs to be explained as a rule. But the advantages of removing score controversy and forcing cleaner/better solutions are pretty strong.

So at this point, that's the only thing to decide: go with it or not.
